src/heap_utils.cc

```diff
@@ -231,19 +235,212 @@ class BufferOutputStream : public v8::OutputStream {
   std::unique_ptr<JSString> buffer_;
 };
 
+namespace {
+class FileOutputStream : public v8::OutputStream {
+ public:
+  explicit FileOutputStream(FILE* stream) : stream_(stream) {}
+
+  virtual int GetChunkSize() {
+    return 65536;  // big chunks == faster
+  }
+
+  virtual void EndOfStream() {}
+
+  virtual WriteResult WriteAsciiChunk(char* data, int size) {
+    const size_t len = static_cast<size_t>(size);
+    size_t off = 0;
+
+    while (off < len && !feof(stream_) && !ferror(stream_))
+      off += fwrite(data + off, 1, len - off, stream_);
+
+    return off == len ? kContinue : kAbort;
+  }
+
+ private:
+  FILE* stream_;
+};
+
+class StreamBaseOutputStream : public AsyncWrap,
+                               public StreamBase,
+                               public v8::OutputStream {
+ public:
+  class StreamBaseOutputStreamShutdownWrap :
+      public ShutdownWrap,
+      public AsyncWrap {
+   public:
+    StreamBaseOutputStreamShutdownWrap(
+      StreamBaseOutputStream* stream,
+        v8::Local<v8::Object> wrap) :
+        ShutdownWrap(stream, wrap),
+        AsyncWrap(stream->env(), wrap, AsyncWrap::PROVIDER_SHUTDOWNWRAP) {}
+
+    AsyncWrap* GetAsyncWrap() override { return this; }
+
+    SET_NO_MEMORY_INFO()
+    SET_MEMORY_INFO_NAME(StreamBaseOutputStreamShutdownWrap)
+    SET_SELF_SIZE(StreamBaseOutputStreamShutdownWrap)
+  };
+
+  StreamBaseOutputStream(
+      Environment* env,
+      const HeapSnapshot* snapshot,
+      v8::Local<v8::Object> obj) :
+      AsyncWrap(env, obj, AsyncWrap::PROVIDER_HEAPDUMP),
+      StreamBase(env),
+      snapshot_(snapshot),
+      ended_(false) {
+    MakeWeak();
+    StreamBase::AttachToObject(GetObject());
+  }
+  ~StreamBaseOutputStream() override {
+    const_cast<HeapSnapshot*>(snapshot_)->Delete();
+  }
+
+  virtual int GetChunkSize() {
+    return 65536;  // big chunks == faster
+  }
+
+  virtual void EndOfStream() {
+    ended_ = true;
+    EmitRead(UV_EOF);
+  }
+
+  virtual WriteResult WriteAsciiChunk(char* data, int size) {
+    uv_buf_t buf = uv_buf_init(data, size);
+    EmitRead(size, buf);
```

Comment:
This doesn’t quite fulfill the `StreamBase` API, because it relies on the memory behind `EmitRead()` having been allocated with `EmitAlloc()` – I think this is okay, but maybe it’s best to leave a comment about it.